### PR TITLE
First implementation

### PR DESCRIPTION
**Pull request by lima-limon-inc**
This creates the `add_children_leaves`

This function creates the first set of children. IN THEORY, this should also work to create n children. I'm still thinking about it.

### CODE REVIEW

File: src/merkle_tree.rs [hunk 1]
```diff
@@ -1,25 +1,115 @@
...
-    let hash = hasher.finalize();
-    println!("Binary hash: {:?}", hash);
+impl MerkleTree {
+    pub fn new<T: std::convert::AsRef<[u8]>>(data: &[T]) -> MerkleTree {
+        let original_leaves: Vec<(Position, Leaf)> = data
+            .iter()
+            .map(|value| {
+                let hash = Sha3_256::digest(value);
+                hash
+            })
```
Comment (LeanSerra) [inline src/merkle_tree.rs]: this can be rewritten as `Sha3_256::digest(value)` if you are returning the hash

File: src/merkle_tree.rs [hunk 2]
```diff
@@ -1,25 +1,115 @@
...
+                // the first element by itself. The second element will be none
+                if a.get(1).is_none() {
+                    [&a[0], &a[0]]
+                } else {
+                    [&a[0], &a[1]]
+                }
+            })
+            .fold(Vec::new(), |mut acc, x| {
+                let positions = [x[0].0, x[1].0];
+                let leaves = [&x[0].1, &x[1].1];
```
Comment (LeanSerra) [inline src/merkle_tree.rs]: x could have a better variable name

File: src/merkle_tree.rs [hunk 3]
```diff
@@ -1,25 +1,115 @@
...
+            .chunks(2)
+            .into_iter()
+            .map(|a| {
+                // This is the case where there is an uneven amount
+                // of data elements. The chunks function will returns
+                // the first element by itself. The second element will be none
+                if a.get(1).is_none() {
+                    [&a[0], &a[0]]
+                } else {
+                    [&a[0], &a[1]]
```
Comment (LeanSerra) [inline src/merkle_tree.rs]: `a` could have a better name

**Review by LeanSerra — APPROVED** [review]
LGTM but i suggested some style changes

File: src/merkle_tree.rs [hunk 4]
```diff
@@ -1,25 +1,115 @@
...
+    pub fn new<T: std::convert::AsRef<[u8]>>(data: &[T]) -> MerkleTree {
+        let original_leaves: Vec<(Position, Leaf)> = data
+            .iter()
+            .map(|value| {
+                let hash = Sha3_256::digest(value);
+                hash
+            })
+            .map(|hash| Leaf::new(hash.into(), None, None))
+            .enumerate()
+            .collect();
```
Comment (damiramirez) [inline src/merkle_tree.rs:65]: I'm not sure what this part does. Maybe we can add comments here?

File: src/merkle_tree.rs (hunk 2 above)
Comment (lima-limon-inc) [reply src/merkle_tree.rs]: Totally fair

File: src/merkle_tree.rs (hunk 1 above)
Comment (lima-limon-inc) [reply src/merkle_tree.rs]: Noted.

**Review by damiramirez — APPROVED** [review]

**Review by LeanSerra — APPROVED** [review]
LGTM

**Review by damiramirez — APPROVED** [review]

**Pull request merged**